### PR TITLE
Prevent double serialization inside Flask server

### PR DESCRIPTION
**Pull request by tdene**
# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: tests/unit_tests/inference/test_data_parallel_inference_coordinator.py [hunk 1]
```diff
@@ -223,7 +223,7 @@ async def run_coordinator_test(
                 results = await asyncio.wait_for(asyncio.gather(*futures), timeout=10.0)
 
                 for record in results:
```
Comment (lmcafee-nvidia) [inline tests/unit_tests/inference/test_data_parallel_inference_coordinator.py]: rename `record` -> `request`

File: megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/chat_completions.py [hunk 2]
```diff
@@ -120,7 +120,7 @@ async def chat_completions():
 
         request_idx = 0
         for record in batch_results:
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/chat_completions.py]: I believe this is a request now, so let's rename `record` -> `request`

File: tests/unit_tests/inference/test_data_parallel_inference_coordinator.py [hunk 3]
```diff
@@ -223,7 +223,7 @@ async def run_coordinator_test(
...
-                    assert record[-1].status == Status.COMPLETED
+                    assert record.status == Status.COMPLETED.name
```
Comment (lmcafee-nvidia) [inline tests/unit_tests/inference/test_data_parallel_inference_coordinator.py]: this file instantiates `InferenceClient` with the default `deserialize = True` as far as I can tell, so does this assert actually work? I would assume it should be `Status.COMPLETED` instead of `Status.COMPLETED.name` unless I'm missing something.

File: megatron/core/inference/inference_client.py [hunk 4]
```diff
@@ -53,13 +53,16 @@ class InferenceClient:
             completed requests.
     """
 
-    def __init__(self, inference_coordinator_address: str):
+    def __init__(self, inference_coordinator_address: str, deserialize: bool = True):
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/inference_client.py]: why are we defaulting to `True` instead of `False`, since the goal is to avoid double serialization?

File: tests/unit_tests/inference/test_data_parallel_inference_coordinator.py
Comment (lmcafee-nvidia) [inline tests/unit_tests/inference/test_data_parallel_inference_coordinator.py:1]: we might want to make sure we have test coverage for both `deserialize=True` and `deserialize=False`, especially since we use `isinstance(result, dict)` to check the object type.

File: examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py [hunk 5]
```diff
@@ -134,7 +134,7 @@ async def main(
             throughputs = []
 
             for record in results:
```
Comment (lmcafee-nvidia) [inline examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py]: record -> request

File: examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py [hunk 6]
```diff
@@ -158,7 +158,7 @@ async def main(
             print("Results:")
             unique_prompt_map = defaultdict(list)
             for record in results:
```
Comment (lmcafee-nvidia) [inline examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py]: record -> request

File: megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/completions.py [hunk 7]
```diff
@@ -129,29 +129,27 @@ async def completions():
 
         request_idx = 0
         for record in batch_results:
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/completions.py]: record -> request

File: megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/completions.py [hunk 8]
```diff
@@ -129,29 +129,27 @@ async def completions():
...
-            result = record.merge()
-            full_text = result.generated_text or ""
+            record_dict = record if isinstance(record, dict) else record.serialize()
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/completions.py]: record_dict -> request_dict

Comment (tdene) [reply megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/completions.py]: Addressed

File: megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/completions.py (hunk 7 above)
Comment (tdene) [reply megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/completions.py]: Addressed

File: examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py (hunk 6 above)
Comment (tdene) [reply examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py]: Addressed

File: examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py (hunk 5 above)
Comment (tdene) [reply examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py]: Addressed

File: tests/unit_tests/inference/test_data_parallel_inference_coordinator.py
Comment (tdene) [reply tests/unit_tests/inference/test_data_parallel_inference_coordinator.py:1]: Addressed

File: megatron/core/inference/inference_client.py (hunk 4 above)
Comment (tdene) [reply megatron/core/inference/inference_client.py]: Addressed

File: tests/unit_tests/inference/test_data_parallel_inference_coordinator.py (hunk 3 above)
Comment (tdene) [reply tests/unit_tests/inference/test_data_parallel_inference_coordinator.py]: Addressed: I changed the default and made sure the whole test file was solid

File: megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/chat_completions.py (hunk 2 above)
Comment (tdene) [reply megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/chat_completions.py]: Addressed

File: tests/unit_tests/inference/test_data_parallel_inference_coordinator.py (hunk 1 above)
Comment (tdene) [reply tests/unit_tests/inference/test_data_parallel_inference_coordinator.py]: Addressed

File: megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/completions.py [hunk 9]
```diff
@@ -128,30 +128,32 @@ async def completions():
...
-            full_text = result.generated_text or ""
+        for completed_request in batch_results:
+            request_dict = (
+                completed_request
+                if isinstance(completed_request, dict)
+                else completed_request.serialize()
+            )
+            # Unwrap ("tensor", [...]) tuples from serialize() into plain lists.
+            result = {
+                k: v[1] if isinstance(v, (list, tuple)) and len(v) == 2 and v[0] == "tensor" else v
```
Comment (santhnm2) [inline megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/completions.py]: This conversion seems like it's potentially error-prone, can we move this to a small helper function like `_unwrap_completed_request` so that it will be a bit easier to debug if necessary?

File: tests/unit_tests/inference/test_data_parallel_inference_coordinator.py [hunk 10]
```diff
@@ -135,10 +135,14 @@ def add_request(
     ) -> asyncio.Future[DynamicInferenceRequestRecord]:
         """Dummy add_request."""
 
+        prompt_tokens = (
+            torch.arange(len(prompt.split())) if isinstance(prompt, str) else torch.tensor(prompt)
+        )
```
Comment (santhnm2) [inline tests/unit_tests/inference/test_data_parallel_inference_coordinator.py:144]: Can you add a comment that you're doing this to mock tokenization? Looked wrong to me at first glance but then I released this was a test file

File: megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/completions.py (hunk 9 above)
Comment (tdene) [reply megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/completions.py]: Addressed!

File: tests/unit_tests/inference/test_data_parallel_inference_coordinator.py (hunk 10 above)
Comment (tdene) [reply tests/unit_tests/inference/test_data_parallel_inference_coordinator.py:144]: Addressed!
